feat(python): show properties and values in TypeBuilder string representation

afyef opened the pull request:
This PR enhances the TypeBuilder's string representation to show all properties of a class and values of an enum, including their aliases and descriptions.

Changes:
- Show all properties in class representation:
  ```python
  # Before:
  TypeBuilder(Classes: ['User'])
  
  # After:
  TypeBuilder(Classes: ['User { name (alias='username', desc='The user\'s full name'), age (desc='User\'s age in years'), email }'])
  ```

- Show all values in enum representation:
  ```python
  # Before:
  TypeBuilder(Enums: ['Status'])
  
  # After:
  TypeBuilder(Enums: ['Status { ACTIVE (alias='active', desc='User is active'), INACTIVE (alias='inactive'), PENDING }'])
  ```

- Store and track metadata:
  - Cache aliases and descriptions in builders
  - Sort properties and values for consistent output
  - Properly escape quotes in string values


Fixes #1253

aaronvg commented:
nice work! We'll have someone in the team review soon

hellovai reviewed (changes requested):
@revidious this is really great work, but if possible, would you be ok implenting this in rust instead?

otherwise we'll have different printers in python/ts/ruby/etc and will make it harder to catch bugs later on (and small details will easily be missed)

afyef committed:
feat: implement string representation for TypeBuilder - Add Display implementation for TypeBuilder and related structs - Add Python bindings for string representation - Add comprehensive test cases for both Rust and Python - Add detailed documentation for string formatting

afyef commented:
@hellovai Done! I've moved the implementation to Rust with the Display trait, and Python now just exposes it via __str__.

hellovai committed:
Merge branch 'canary' into feature/typebuilder-repr

hellovai reviewed (changes requested):
this is great! :) ready to merge. small request, could you add this for the language_client_{ts / ruby} as well. If not, let us know and i'll follow up on it instead. This is almost perfect btw 👍 

Some nits things I would do (but not mandatory): 
* add whitespace / newlines in the rendered string
* instead of `BamlValue::to_string(...)` I would just implement the display for all baml values
* Instead of hard-coding which attributes i print out for meta, i would inject all metas in the print out, not just alias / description

afyef committed:
feat(typebuilder): add typescript wrapper and tests for type builder

afyef committed:
docs(ruby): add detailed comments to typebuilder rust and ruby test files

afyef committed:
feat(typebuilder): display all metadata fields in string representation

afyef committed:
feat(typebuilder): improve string representation formatting and display all metadata fields

afyef commented:
Thanks for the feedback! I've implemented all three suggestions:
- Added proper whitespace/newlines in the rendered string
- Using BamlValue's Display trait directly instead of to_string()
- Now showing all metadata fields without specific ones

Changes are implemented in both TypeScript and Ruby clients as well (what it seems like).

aaronvg commented:
It seems there's some issues with:
"FAILED test_type_builder_str.py::test_type_builder_str - AttributeError: 'TypeBuilder' object has no attribute 'class_'"

If you can take a look, it seems this PR is close

hellovai commented:
hey @revidious! Just wanted to follow up here! Glad to fork this PR and make the fixes myself if you are busy. this is a great one and we'd love to merge this in!

hellovai reviewed (dismissed):
waiting for tests to pass

hellovai committed:
Merge branch 'canary' into feature/typebuilder-repr

afyef commented:
Hey @hellovai! Thanks for following up. I’ve been a bit caught up with life and other things, but I really appreciate your offer to fork the PR and make the fixes yourself. I’ll try to get the tests sorted and the changes addressed soon. Thanks so much for your patience!

hellovai committed:
update tests

hellovai committed:
Merge branch 'canary' into feature/typebuilder-repr

hellovai committed:
udpate lock file

aaronvg committed:
add python integ test

aaronvg committed:
Merge branch 'canary' into feature/typebuilder-repr

aaronvg committed:
Add typebuilder string representation

aaronvg committed:
delete extraneous test

aaronvg commented:
@afyef amazing work! Got this one merged for you
